File: exampleCourse/elementExtensions/pl-drawing/example-logo/example-logo.js
class PrairieLearnLogo extends PLDrawingBaseElement {
  static generate(canvas, options, submittedAnswer) {
    fabric.Image.fromURL(options.image_url, (obj) => {
      /* Generate a unique ID for this object if it doesn't have one */
      if (!('id' in obj)) {
        obj.id = PLDrawingApi.generateID();
      }

      /* Set the Fabric object's values from what we loaded */
      obj.set({
        left: options.left,
        top: options.top,
        angle: options.angle,
        originX: 'center',
        originY: 'center',
      });

      /* Disable the scaling controls */
      obj.setControlsVisibility({
        mt: false,
        mb: false,
        ml: false,
        mr: false,
        bl: false,
        br: false,
        tl: false,
        tr: false,
      });

      canvas.add(obj);
      submittedAnswer.registerAnswerObject(options, obj);
    });
  }

  static get_button_tooltip() {
    return 'Create PrairieLearn logo';
  }

  static get_button_icon() {
    return 'logo-icon.svg';
  }
}

PLDrawingApi.registerElements('example-logo', {
  'pl-prairielearn-logo': PrairieLearnLogo,
});

File: exampleCourse/elementExtensions/pl-drawing/example-logo/example-logo.py
from os import path

import prairielearn as pl

# Import class definitions and default values from the drawing element
defaults = pl.load_host_script("defaults.py")
elements = pl.load_host_script("elements.py")


# Elements
class PrairieLearnLogo(elements.BaseElement):
    def generate(el, data):
        return {
            "left": pl.get_float_attrib(el, "x", 20),
            "top": pl.get_float_attrib(el, "y", 20),
            "angle": pl.get_float_attrib(el, "angle", 0),
            "image_url": path.join(data["clientFilesUrl"], "logo.png"),
        }

    def get_attributes():
        return ["x", "y", "angle"]


elements = {}
elements["pl-prairielearn-logo"] = PrairieLearnLogo
